godot-core/src/init/mod.rs

```diff
@@ -120,27 +122,31 @@ unsafe extern "C" fn ffi_deinitialize_layer<E: ExtensionLibrary>(
 }
 
 /// Tasks needed to be done by gdext internally upon loading an initialization level. Called before user code.
-fn gdext_on_level_init(level: InitLevel) {
-    // SAFETY: we are in the main thread, during initialization, no other logic is happening.
+///
+/// # Safety
+///
+/// - Must be called from the main thread.
+/// - The interface must have been initialized.
+/// - Must only be called once per level.
+#[deny(unsafe_op_in_unsafe_fn)]
+unsafe fn gdext_on_level_init(level: InitLevel) {
     // TODO: in theory, a user could start a thread in one of the early levels, and run concurrent code that messes with the global state
     // (e.g. class registration). This would break the assumption that the load_class_method_table() calls are exclusive.
     // We could maybe protect globals with a mutex until initialization is complete, and then move it to a directly-accessible, read-only static.
-    unsafe {
-        match level {
-            InitLevel::Core => {}
-            InitLevel::Servers => {
-                sys::load_class_method_table(sys::ClassApiLevel::Server);
-            }
-            InitLevel::Scene => {
-                sys::load_class_method_table(sys::ClassApiLevel::Scene);
-                ensure_godot_features_compatible();
-            }
-            InitLevel::Editor => {
-                sys::load_class_method_table(sys::ClassApiLevel::Editor);
-            }
-        }
-        crate::registry::class::auto_register_classes(level);
+    let Some(class_api_level) = level.to_class_api_level() else {
+        // `level` is core and no initialization is needed during `Core`.
+        return;
+    };
+
+    // SAFETY: we are in the main thread, initialize has been called, has never been called with this level before.
+    unsafe { sys::load_class_method_table(class_api_level) };
+
+    if level == InitLevel::Scene {
+        // SAFETY: On the main thread, api initialized, `Scene` was initialized above.
+        unsafe { ensure_godot_features_compatible() };
     }
+
+    crate::registry::class::auto_register_classes(level);
```

Comment:
Previously there was a single match that would immediately show the behavior for all 4 levels.

Now we have the case differentiation in two places (here and `to_class_api_level`), and there are lots of extra relations like "None means Core" and "if Scene then handle special case".

I don't like making code flow more complex to satisfy `unsafe` (similar problem as in https://github.com/godot-rust/gdext/pull/741#discussion_r1626151678). The previous flow was trivial to understand and didn't require an extra function.

Reply:
i honestly didnt find it much simpler to understand. That's a large part of why i changed it tbh. This code does show it fairly straightforward imo:

1. Do nothing for core
2. load class method table
3. When loading scene, ensure the godot features are compatible
4. Auto register this level's classes

Whereas the previous one had 4 branches, where 3 of them did the same thing except one of them had an extra condition. So it became more like
1. check api level
  - for core do nothing
  - for servers, load servers method table
  - for scene load scene method table and ensure godot features compatible
  - for editor load editor method table
2. auto register classes for current level

Personally i find the first one clearer since it's more linear, and less branching.

`None` meaning core could also be rephrased as "`None` means we cannot load the class method table for this api level". Since really the only reason it does return `Option` is that `load_class_method_table` can only handle `Servers`, `Scene`, and `Editor`.

Reply:
"Easy to understand" is very subjective, so I'll try to be more precise what I meant:

The previous code flow was linear. If I'm interested in what `Scene` level does, I can start from top to bottom and see it loads the table, ensures compatibility and then registers classes. 

With your approach, I first have to jump into `to_class_api_level`, read the mapping, see that it is `Some` and thus does not hit the `return`. Then, it loads the table and there's another condition `if level == InitLevel::Scene`, which is true, i.e. the inner block is also executed. There is non-local data flow that influences which branches are taken, I really can't see how this is simpler.

---

> Whereas the previous one had 4 branches, where 3 of them did the same thing except one of them had an extra condition.

The fact that 3 did the same thing is immediately visible. We don't have to forcefully deduplicate `sys::load_class_method_table()` calls.

Reply:
> "Easy to understand" is very subjective, so I'll try to be more precise what I meant:
> 
> The previous code flow was linear. If I'm interested in what `Scene` level does, I can start from top to bottom and see it loads the table, ensures compatibility and then registers classes.
> 
> With your approach, I first have to jump into `to_class_api_level`, read the mapping, see that it is `Some` and thus does not hit the `return`. Then, it loads the table and there's another condition `if level == InitLevel::Scene`, which is true, i.e. the inner block is also executed. There is non-local data flow that influences which branches are taken, I really can't see how this is simpler.

I think this is largely a flaw of us having different types that represent the same thing tbh. Because `InitLevel` and `ClassApiLevel` are basically the same information. Except that `ClassApiLevel` can't be `Core`. So when i look at that and know the types, i know that `to_class_api_level` should just a direct mapping of `InitLevel` to `ClassApiLevel` and the only reason to check the method is to verify that it is implemented correctly.

But i dont see why we couldn't just merge them into one and just have `Core` do nothing (or in the future maybe it can do something if we allow registering at that level) in `load_class_method_table`. 

> > Whereas the previous one had 4 branches, where 3 of them did the same thing except one of them had an extra condition.
> 
> The fact that 3 did the same thing is immediately visible. We don't have to forcefully deduplicate `sys::load_class_method_table()` calls.

Sorta? To me it did lead to me checking to ensure that it did what i expected. Like i needed to go through each and see that each one just registered with the same level that was declared in the match branch.

Reply:
Good observation, maybe we should merge the two enums then?

I'm fine with `load_class_method_table` asserting that it isn't accidentally invoked with `Core`.

Reply:
I merged `InitLevel` and `ApiLevel` into one, and i chose `ApiLevel` as the single name for that. I chose `ApiLevel` cause it's a more accurate name imo, regardless of the context it's used in it always describes the godot api level of something. When used as an init level, it is which api level it is initialized for. This on its own would technically be a major breaking change, so i added `InitLevel` as a deprecated type alias for `ApiLevel`. Also the two disagreed on what the server level was called, in `ApiLevel` it was `Server` and in `InitLevel` it was `Servers`. So i made them both just be `Server` with a deprecated pub const `Servers`. All this should be only a minor change.

Currently i just made `load_class_method_table` do nothing when called with `Core`. I could also make it assert that it's not called with `Core` but that would mean that we need to check upfront that it's not, and i think it's simpler to just make it do nothing.